pandas/core/arrays/sparse/accessor.py

```diff
@@ -313,7 +331,7 @@ def from_spmatrix(cls, data, index=None, columns=None) -> DataFrame:
         indices = data.indices
         indptr = data.indptr
         array_data = data.data
-        dtype = SparseDtype(array_data.dtype, 0)
+        dtype = SparseDtype(array_data.dtype, fill_value)
```

Comment:
Can you use `na_value_for_dtype` instead of introducing a new argument?

Reply:
Hi Matthew. Thank you for your speedy response and taking the time to review my PR.

``na_value_for_dtype`` is indirectly called in this implementation with the default argument of ``fill_value`` as ``None`` when passed to the ``SparseDtype`` constructor.

https://github.com/pandas-dev/pandas/blob/a5e812d86deb62872f8d514d894a22931fc84217/pandas/core/dtypes/dtypes.py#L1746-L1747

A ``fill_value`` parameter is typically only needed when constructing a sparse format object from a dense format object, as we need to identify non-zero elements in the data to set attributes like the data, index, and index pointer (for sparse array formats like BSR, CSR, and CSC) in the sparse format object correctly. In this case, we are simply accessing these attributes from a CSC matrix, so you are right in that a ``fill_value`` parameter is not strictly required to solve the bug.

However, in addition to fixing the bug and not adding overhead, it is handy to have a ``fill_value`` parameter to give the user flexibility in certain use cases, for example, converting a ``SparseArray`` to a ``np.ndarray`` using ``np.asarray`` will use ``fill_value`` to populate the missing elements. This is along the lines of how, albeit not a sparse implementation, ``np.ma.core.MaskedArray`` has a ``filled`` method that performs the same functionality of using a custom ``fill_value`` to convert to a ``np.ndarray``.

Would you be OK with me keeping it?

The current test failures also seem unrelated to the PR and have been around since #59027.

Reply:
The core maintainers have been cautious about expanding the APIs surface areas and signatures unless necessary or for consistency issues. I would prefer if this bug was solved without adding a new keyword argument first, and then you could open a new issue about adding a new keyword here that has opt-in from more core team memebers

Reply:
I understand and yes, the cautious approach makes total sense given the size of pandas.

The issues that I see arising as a result of removing the ``fill_value`` parameter and implementing your change are:

- ``DataFrame.sparse.from_spmatrix().sparse.to_coo``, shown as an example in the user guide, will break for ``float`` and ``complex`` subtypes as it will raise a ``ValueError``. It seems like the original thought of adding this was because a custom ``fill_value`` will be lost upon converting to a COO matrix as ``scipy.sparse._coo.coo_matrix`` and other sparse formats do not have an analogous attribute and instead use ``False``, ``0``, ``0.``, and ``0. + 0.j`` when returning a dense representation of them as a ``np.ndarray`` or ``np.matrix``, which was considered unexpected behaviour at the time. :shrug: We can remove this check as the constructor called in ``to_coo`` is using the _ijv_ format and not directly instantiating using a single 2-D ``np.ndarray``, so the COO matrix returned will be correct regardless of the ``fill_value``. https://github.com/pandas-dev/pandas/blob/b1e5f0636ab2cc3e6949a051e37dc289616a4a4f/doc/source/user_guide/sparse.rst?plain=1#L191-L200
https://github.com/pandas-dev/pandas/blob/b1e5f0636ab2cc3e6949a051e37dc289616a4a4f/pandas/core/arrays/sparse/accessor.py#L396-L397
- All of the tests that rely on a ``DataFrame.sparse.from_spmatrix`` invocation, apart from the one that I added to test the changes, ``test_from_spmatrix_fill_value``, assume a ``fill_value`` of `0`. The expectations will all have to be changed and will subsequently have a dependency on ``na_value_for_dtype`` for the ``fill_value`` to use, and the docstring example will have to change.

Would you like me to make the changes above, close the PR, or keep the ``fill_value`` parameter as it is currently implemented in my branch?

Reply:
I think the changes above would be appropriate

Reply:
Excellent. It should be rather straightforward and quick in that case! I will make the changes now.